Bottom: a7a4d704405587ab2550a23dca9483d54f50b99e
Top:    55eb0f7099d8b38fb4fe3fee89d24ff831e2b232
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-09-12 01:25:02 +0100

Revert "We can just delete kicked out holes"

This reverts commit 6400e974907264b9a0524a0dbdd7dfb9eb623f8b.

We can't (although I'm not 100% sure why).

Basically, deleting a series of polygons which creates a hole
in a certain way causes problems with this commit.
 __________
|    ___   |
|   |___|  |
|__________|

These lines represent line polygons which are deleted one by one
from a larger "world" polygon in the GL version's mask rendering.

The order is such that three of the outer polygons are removed first,
leaving a mirrored C shaped hole in the world polygon:
 __________
           |
           |
 __________|

When subtracing some of the inner polygons, they are erronously
discarded with this patch applied.


---

diff --git a/src/polygon1.c b/src/polygon1.c
index 4a761d5..e7642c9 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -1423,8 +1423,7 @@ InsertHoles (jmp_buf * e, POLYAREA * dest, PLINE ** src)
 	}
       else
 	{
-	  /* Need to check if this new hole means we need to kick out any old
-	   * ones for deletion */
+	  /* Need to check if this new hole means we need to kick out any old ones for reprocessing */
 	  while (1)
 	    {
 	      struct find_inside_info info;
@@ -1458,7 +1457,10 @@ InsertHoles (jmp_buf * e, POLYAREA * dest, PLINE ** src)
 
 	      /* Remove hole from the contour */
 	      remove_contour (pa_info->pa, prev, info.result, TRUE);
-	      poly_DelContour (&info.result);
+
+	      /* Add hole as the next on the list to be processed in this very function */
+	      info.result->next = *src;
+	      *src = info.result;
 	    }
 	  /* End check for kicked out holes */
